parlai/agents/examples/tra.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+This is an example how to extend torch ranker agent and use it for your own purpose.
+In this example, we will just use a simple bag of words model.
+"""
+from parlai.core.torch_ranker_agent import TorchRankerAgent
+import torch
+from torch import nn
+
+
+class ExampleBagOfWordsModel(nn.Module):
+    """
+    This constructs a simple bag of words model. 
+    It contains a encoder for encoding candidates and context.
+    """
+
+    def __init__(self, opt, dictionary):
+        super().__init__()
+        self.hidden_dim = opt.get('hidden_dim', 512)
+        self.dict = dictionary
+        self.encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+
+    def forward_candidates(self, cand_vecs):
+        """
+        This function will encode all the candidates for a batch. It can be different cases:
+        1. inline candidate: we have inline candidates provided for each sample, it usually happens during eval / test time
+        In this case, the cand_vecs is a 3D tensor, [btz, num_candidates, seq_length]
+        2. batch candidates, it usually happens during training, cand_vecs will be [btz, seq_length]
+        3. fixed, when you have a fixed set of candidates to use, cand_vecs will be [num_candidates, seq_length]
+        """
+        return self.encoder.forward(cand_vecs)
```

Comment:
also why do you explicitly need to call forward here?  if this function is the same for the context/candidates, then why do we need separate functions? just for clarity?

Reply:
Yes, I just did it for clarity purposes. I will remove it then!